src/util/rust_util/sos_util.rs

```diff
@@ -0,0 +1,274 @@
+    /// -   `get_key`: a closure that maps each element to a key.
+    pub fn new(slice_of_slices: &'s [&'s [T]], get_key: F) -> Self {
+        Self {
+            slice_of_slices,
+            get_key,
+            start_outer: 0,
+            start_inner: 0,
+            start_total: 0,
+            end_outer: 0,
+            end_inner: 0,
+            end_total: 0,
+            maybe_prev_key: None,
+            done: false,
+        }
+    }
+
+    fn inspect_current_element(&mut self) -> (K, Option<Group<'s, T, K>>) {
+        debug_assert!(
+            !self.slice_of_slices[self.end_outer].is_empty(),
+            "An inner slice is empty"
+        );
+
+        let cur_key = (self.get_key)(&self.slice_of_slices[self.end_outer][self.end_inner]);
+
+        let maybe_group = match self.maybe_prev_key {
+            Some(prev_key) if prev_key != cur_key => Some(self.capture()),
+            _ => None,
+        };
+
+        (cur_key, maybe_group)
+    }
+
+    fn capture(&self) -> Group<'s, T, K> {
+        debug_assert!(self.maybe_prev_key.is_some());
+
+        Group {
+            slice_of_slices: self.slice_of_slices,
+            key: self.maybe_prev_key.unwrap(),
+            start_outer: self.start_outer,
+            start_inner: self.start_inner,
+            start_total: self.start_total,
+            end_outer: self.end_outer,
+            end_inner: self.end_inner,
+            end_total: self.end_total,
+        }
+    }
+
+    fn has_next(&self) -> bool {
+        self.end_outer < self.slice_of_slices.len()
+    }
+
+    fn reset_start(&mut self) {
+        self.start_outer = self.end_outer;
+        self.start_inner = self.end_inner;
+        self.start_total = self.end_total;
+    }
+
+    fn proceed(&mut self) {
+        self.end_total += 1;
+        self.end_inner += 1;
+        if self.end_inner == self.slice_of_slices[self.end_outer].len() {
+            self.end_inner = 0;
+            self.end_outer += 1;
+        }
+    }
+}
+
+impl<'s, T, K, F> Iterator for SliceOfSlicesGrouper<'s, T, K, F>
+where
+    K: PartialEq + Copy,
+    F: FnMut(&T) -> K,
+{
+    type Item = Group<'s, T, K>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.has_next() {
+            let (cur_key, maybe_group) = self.inspect_current_element();
+            self.maybe_prev_key = Some(cur_key);
+            if maybe_group.is_some() {
+                self.reset_start();
+            }
+            self.proceed();
+
+            if maybe_group.is_some() {
+                return maybe_group;
+            }
+        }
+
+        if self.maybe_prev_key.is_some() {
+            debug_assert_eq!(self.end_outer, self.slice_of_slices.len());
+            debug_assert_eq!(self.end_inner, 0);
+
+            let group = self.capture();
+            self.maybe_prev_key = None;
+
+            Some(group)
+        } else {
+            None
+        }
+    }
+}
+
+#[test]
```

Comment:
Probably put the tests to a `#[cfg(test)]` module to avoid including them in the normal generated binary.

Reply:
Done. I moved tests into a `tests` module.